server/bin/test_data_importer.py

```diff
@@ -0,0 +1,706 @@
+        }
+
+        # Merge in the other keys we care about for pools
+        keys = ['branding', 'contractNumber', 'accountNumber', 'orderNumber', 'subscriptionId', 'upstreamPoolId']
+        safe_merge(pool_data, pool, include=keys)
+
+        if self.hosted_test:
+            # Need to set a subscription ID if we're generating this upstream
+            pool['id'] = random_id('upstream')
+
+            # Product is an object in hosted test mode
+            pool['product'] = { 'id': pool_data['product_id'] }
+
+            # The owner must be set in the subscription data for upstream subscriptions
+            pool['owner'] = { 'key': owner_key }
+
+            url = 'hostedtest/subscriptions'
+        else:
+            # Copy over the pool (special case because hosted test data is different)
+            pool['productId'] = pool_data['product_id']
+
+            # Generate fields that make this look like an upstream subscription
+            if not 'subscriptionId' in pool:
+                pool['subscriptionId'] = random_id('srcsub')
+
+            if 'subscription_subkey' in pool_data:
+                pool['subscriptionSubKey'] = pool_data['subscription_subkey']
+            elif 'subscriptionSubKey' in pool_data:
+                pool['subscriptionSubKey'] = pool_data['subscriptionSubKey']
+            else:
+                pool['subscriptionSubKey'] = 'master'
+
+            if not 'upstreamPoolId' in pool:
+                pool['upstreamPoolId'] = random_id('upstream')
+
+            url = "owners/%s/pools" % (owner_key)
+
+        return self.post(url, pool).json()
+
+    def list_pools(self, owner_key):
+        return self.get("owners/%s/pools" % (owner_key)).json()
+
+    def create_activation_key(self, owner_key, key_data):
+        key = safe_merge(key_data, {})
+        return self.post("owners/%s/activation_keys" % (owner_key), key).json()
+
+    def add_pool_to_activation_key(self, key_id, pool_id, quantity=None):
+        url = "activation_keys/%s/pools/%s" % (key_id, pool_id)
+        params = {}
+
+        if quantity is not None:
+            params['quantity'] = quantity
+
+        return self.post(url, None, params).json()
+
+    def refresh_pools(self, owner_key, wait_for_completion=True):
+        job_status = self.put("owners/%s/subscriptions" % (owner_key), None, params={ 'lazy_regen': True }).json()
+
+        if job_status and wait_for_completion:
+            finished_states = ['ABORTED', 'FINISHED', 'CANCELED', 'FAILED']
+
+            while job_status['state'] not in finished_states:
+                time.sleep(1)
+                job_status = self.get(job_status['statusPath']).json()
+
+            # Clean up job status
+            # self.delete('jobs', {'id': job_status['id']})
+
+        return job_status
+
+
+
+# Gather and order content in a way that ensures linear creation is safe
+def gather_content(data, hosted_test):
+    content = []
```

Comment:
One more question. Why is list of tuples used here? You create map (dictionary) of contents later, where key for content is content ID. What about duplicity in generated list? I would expect more comments and explanation in this function.

Reply:
The tuples here and the map you noticed later have different sets of data and serve different purposes:

The output of gather_content is an array of tuples consisting of the content ID and the key of the owning organization (if applicable). In standalone/Sat mode, the content ID may be present multiple times, as each org gets their own copy of the content; whereas in hosted/portal mode, the content is created globally, and will not be assigned to an org. In both cases it's technically incorrect to have a given tuple of (content_id, owner_key) appear more than once, so I suppose we could use a set here to eliminate that potential risk (assuming set is py2 compatible); however, we still don't validate across files, so the test data still needs to be formatted properly to avoid the headache here.

The output of map_content_to_product (which is what I assume you're referencing) generates the mapping of content_id to a boolean indicating whether or not that content is enabled. This data comes from a different location (product['content']), and only overlaps with data['content'] and data['owner']['content'] in the content_id. This map is passed to Candlepin to indicate which content is mapped to a given product, and is built for each product that specifies content.